File: mdinfofiles/bst/BST.php
<?php

/**
 * Node class - represents a single node in the BST
 */
class Node
{
    public int $data;
    public ?Node $left;
    public ?Node $right;

    public function __construct(int $data)
    {
        $this->data = $data;
        $this->left = null;
        $this->right = null;
    }
}

/**
 * BinarySearchTree class - full BST implementation
 */
class BinarySearchTree
{
    private ?Node $root;

    public function __construct()
    {
        $this->root = null;
    }

    // ---------- CREATE / INSERT ----------

    public function insert(int $data): void
    {
        $this->root = $this->insertNode($this->root, $data);
    }

    private function insertNode(?Node $node, int $data): Node
    {
        if ($node === null) {
            return new Node($data);
        }

        if ($data < $node->data) {
            $node->left = $this->insertNode($node->left, $data);
        } elseif ($data > $node->data) {
            $node->right = $this->insertNode($node->right, $data);
        }
        // duplicate values are ignored

        return $node;
    }

    // ---------- SEARCH ----------

    public function search(int $data): bool
    {
        return $this->searchNode($this->root, $data);
    }

    private function searchNode(?Node $node, int $data): bool
    {
        if ($node === null) {
            return false;
        }

        if ($data === $node->data) {
            return true;
        }

        return $data < $node->data
            ? $this->searchNode($node->left, $data)
            : $this->searchNode($node->right, $data);
    }

    // ---------- DELETE ----------

    public function delete(int $data): void
    {
        $this->root = $this->deleteNode($this->root, $data);
    }

    private function deleteNode(?Node $node, int $data): ?Node
    {
        if ($node === null) {
            return null;
        }

        if ($data < $node->data) {
            $node->left = $this->deleteNode($node->left, $data);
        } elseif ($data > $node->data) {
            $node->right = $this->deleteNode($node->right, $data);
        } else {
            // Node found

            // Case 1: no children
            if ($node->left === null && $node->right === null) {
                return null;
            }

            // Case 2: one child
            if ($node->left === null) {
                return $node->right;
            }
            if ($node->right === null) {
                return $node->left;
            }

            // Case 3: two children
            // Find inorder successor (smallest in right subtree)
            $successor = $this->findMin($node->right);
            $node->data = $successor->data;
            $node->right = $this->deleteNode($node->right, $successor->data);
        }

        return $node;
    }

    private function findMin(Node $node): Node
    {
        while ($node->left !== null) {
            $node = $node->left;
        }
        return $node;
    }

    private function findMax(Node $node): Node
    {
        while ($node->right !== null) {
            $node = $node->right;
        }
        return $node;
    }

    // ---------- MIN / MAX ----------

    public function getMin(): ?int
    {
        if ($this->root === null) {
            return null;
        }
        return $this->findMin($this->root)->data;
    }

    public function getMax(): ?int
    {
        if ($this->root === null) {
            return null;
        }
        return $this->findMax($this->root)->data;
    }

    // ---------- HEIGHT ----------

    public function height(): int
    {
        return $this->heightNode($this->root);
    }

    private function heightNode(?Node $node): int
    {
        if ($node === null) {
            return -1; // empty tree height = -1, single node = 0
        }
        return 1 + max($this->heightNode($node->left), $this->heightNode($node->right));
    }

    // ---------- COUNT NODES ----------

    public function countNodes(): int
    {
        return $this->countNodesHelper($this->root);
    }

    private function countNodesHelper(?Node $node): int
    {
        if ($node === null) {
            return 0;
        }
        return 1 + $this->countNodesHelper($node->left) + $this->countNodesHelper($node->right);
    }

    // ---------- TRAVERSALS (SHOW) ----------

    public function inorder(): array
    {
        $result = [];
        $this->inorderHelper($this->root, $result);
        return $result;
    }

    private function inorderHelper(?Node $node, array &$result): void
    {
        if ($node === null) {
            return;
        }
        $this->inorderHelper($node->left, $result);
        $result[] = $node->data;
        $this->inorderHelper($node->right, $result);
    }

    public function preorder(): array
    {
        $result = [];
        $this->preorderHelper($this->root, $result);
        return $result;
    }

    private function preorderHelper(?Node $node, array &$result): void
    {
        if ($node === null) {
            return;
        }
        $result[] = $node->data;
        $this->preorderHelper($node->left, $result);
        $this->preorderHelper($node->right, $result);
    }

    public function postorder(): array
    {
        $result = [];
        $this->postorderHelper($this->root, $result);
        return $result;
    }

    private function postorderHelper(?Node $node, array &$result): void
    {
        if ($node === null) {
            return;
        }
        $this->postorderHelper($node->left, $result);
        $this->postorderHelper($node->right, $result);
        $result[] = $node->data;
    }

    public function levelOrder(): array
    {
        $result = [];
        if ($this->root === null) {
            return $result;
        }

        $queue = [$this->root];
        while (!empty($queue)) {
            $node = array_shift($queue);
            $result[] = $node->data;

            if ($node->left !== null) {
                $queue[] = $node->left;
            }
            if ($node->right !== null) {
                $queue[] = $node->right;
            }
        }

        return $result;
    }

    // ---------- VISUAL DISPLAY ----------

    public function display(): void
    {
        if ($this->root === null) {
            echo "Tree is empty\n";
            return;
        }
        $this->displayHelper($this->root, "", true);
    }

    private function displayHelper(?Node $node, string $prefix, bool $isTail): void
    {
        if ($node === null) {
            return;
        }

        if ($node->right !== null) {
            $this->displayHelper($node->right, $prefix . ($isTail ? "│   " : "    "), false);
        }

        echo $prefix . ($isTail ? "└── " : "┌── ") . $node->data . "\n";

        if ($node->left !== null) {
            $this->displayHelper($node->left, $prefix . ($isTail ? "    " : "│   "), true);
        }
    }

    // ---------- IS EMPTY ----------

    public function isEmpty(): bool
    {
        return $this->root === null;
    }
}

// =====================================================
// DEMO
// =====================================================

$bst = new BinarySearchTree();

$values = [50, 30, 70, 20, 40, 60, 80, 10, 25];
foreach ($values as $v) {
    $bst->insert($v);
}

echo "Tree structure:\n";
$bst->display();

echo "\nInorder (sorted):   " . implode(", ", $bst->inorder()) . "\n";
echo "Preorder:           " . implode(", ", $bst->preorder()) . "\n";
echo "Postorder:          " . implode(", ", $bst->postorder()) . "\n";
echo "Level order (BFS):  " . implode(", ", $bst->levelOrder()) . "\n";

echo "\nMin value: " . $bst->getMin() . "\n";
echo "Max value: " . $bst->getMax() . "\n";
echo "Height: " . $bst->height() . "\n";
echo "Total nodes: " . $bst->countNodes() . "\n";

echo "\nSearch 40: " . ($bst->search(40) ? "Found" : "Not found") . "\n";
echo "Search 99: " . ($bst->search(99) ? "Found" : "Not found") . "\n";

echo "\nDeleting 30 (node with two children)...\n";
$bst->delete(30);
$bst->display();
echo "Inorder after delete: " . implode(", ", $bst->inorder()) . "\n";

echo "\nDeleting 20 (node with one child)...\n";
$bst->delete(20);
$bst->display();

echo "\nDeleting 80 (leaf node)...\n";
$bst->delete(80);
$bst->display();
echo "Inorder after deletes: " . implode(", ", $bst->inorder()) . "\n";

File: mdinfofiles/bst/bst.py
from collections import deque
from typing import Optional, List


class Node:
    """Represents a single node in the BST."""

    def __init__(self, data: int):
        self.data = data
        self.left: Optional["Node"] = None
        self.right: Optional["Node"] = None


class BinarySearchTree:
    """Full BST implementation with create, insert, delete, search, traversals."""

    def __init__(self):
        self.root: Optional[Node] = None

    # ---------- CREATE / INSERT ----------

    def insert(self, data: int) -> None:
        self.root = self._insert_node(self.root, data)

    def _insert_node(self, node: Optional[Node], data: int) -> Node:
        if node is None:
            return Node(data)

        if data < node.data:
            node.left = self._insert_node(node.left, data)
        elif data > node.data:
            node.right = self._insert_node(node.right, data)
        # duplicates ignored

        return node

    # ---------- SEARCH ----------

    def search(self, data: int) -> bool:
        return self._search_node(self.root, data)

    def _search_node(self, node: Optional[Node], data: int) -> bool:
        if node is None:
            return False
        if data == node.data:
            return True
        return self._search_node(node.left if data < node.data else node.right, data)

    # ---------- DELETE ----------

    def delete(self, data: int) -> None:
        self.root = self._delete_node(self.root, data)

    def _delete_node(self, node: Optional[Node], data: int) -> Optional[Node]:
        if node is None:
            return None

        if data < node.data:
            node.left = self._delete_node(node.left, data)
        elif data > node.data:
            node.right = self._delete_node(node.right, data)
        else:
            # Node found

            # Case 1: no children
            if node.left is None and node.right is None:
                return None

            # Case 2: one child
            if node.left is None:
                return node.right
            if node.right is None:
                return node.left

            # Case 3: two children -> inorder successor
            successor = self._find_min(node.right)
            node.data = successor.data
            node.right = self._delete_node(node.right, successor.data)

        return node

    def _find_min(self, node: Node) -> Node:
        while node.left is not None:
            node = node.left
        return node

    def _find_max(self, node: Node) -> Node:
        while node.right is not None:
            node = node.right
        return node

    # ---------- MIN / MAX ----------

    def get_min(self) -> Optional[int]:
        if self.root is None:
            return None
        return self._find_min(self.root).data

    def get_max(self) -> Optional[int]:
        if self.root is None:
            return None
        return self._find_max(self.root).data

    # ---------- HEIGHT ----------

    def height(self) -> int:
        return self._height_node(self.root)

    def _height_node(self, node: Optional[Node]) -> int:
        if node is None:
            return -1  # empty tree height = -1, single node = 0
        return 1 + max(self._height_node(node.left), self._height_node(node.right))

    # ---------- COUNT NODES ----------

    def count_nodes(self) -> int:
        return self._count_helper(self.root)

    def _count_helper(self, node: Optional[Node]) -> int:
        if node is None:
            return 0
        return 1 + self._count_helper(node.left) + self._count_helper(node.right)

    # ---------- TRAVERSALS (SHOW) ----------

    def inorder(self) -> List[int]:
        result: List[int] = []
        self._inorder_helper(self.root, result)
        return result

    def _inorder_helper(self, node: Optional[Node], result: List[int]) -> None:
        if node is None:
            return
        self._inorder_helper(node.left, result)
        result.append(node.data)
        self._inorder_helper(node.right, result)

    def preorder(self) -> List[int]:
        result: List[int] = []
        self._preorder_helper(self.root, result)
        return result

    def _preorder_helper(self, node: Optional[Node], result: List[int]) -> None:
        if node is None:
            return
        result.append(node.data)
        self._preorder_helper(node.left, result)
        self._preorder_helper(node.right, result)

    def postorder(self) -> List[int]:
        result: List[int] = []
        self._postorder_helper(self.root, result)
        return result

    def _postorder_helper(self, node: Optional[Node], result: List[int]) -> None:
        if node is None:
            return
        self._postorder_helper(node.left, result)
        self._postorder_helper(node.right, result)
        result.append(node.data)

    def level_order(self) -> List[int]:
        result: List[int] = []
        if self.root is None:
            return result

        queue = deque([self.root])
        while queue:
            node = queue.popleft()
            result.append(node.data)
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)

        return result

    # ---------- VISUAL DISPLAY ----------

    def display(self) -> None:
        if self.root is None:
            print("Tree is empty")
            return
        self._display_helper(self.root, "", True)

    def _display_helper(self, node: Optional[Node], prefix: str, is_tail: bool) -> None:
        if node is None:
            return

        if node.right is not None:
            self._display_helper(node.right, prefix + ("│   " if is_tail else "    "), False)

        print(prefix + ("└── " if is_tail else "┌── ") + str(node.data))

        if node.left is not None:
            self._display_helper(node.left, prefix + ("    " if is_tail else "│   "), True)

    # ---------- IS EMPTY ----------

    def is_empty(self) -> bool:
        return self.root is None


# =====================================================
# DEMO
# =====================================================

if __name__ == "__main__":
    bst = BinarySearchTree()

    values = [50, 30, 70, 20, 40, 60, 80, 10, 25]
    for v in values:
        bst.insert(v)

    print("Tree structure:")
    bst.display()

    print(f"\nInorder (sorted):   {', '.join(map(str, bst.inorder()))}")
    print(f"Preorder:           {', '.join(map(str, bst.preorder()))}")
    print(f"Postorder:          {', '.join(map(str, bst.postorder()))}")
    print(f"Level order (BFS):  {', '.join(map(str, bst.level_order()))}")

    print(f"\nMin value: {bst.get_min()}")
    print(f"Max value: {bst.get_max()}")
    print(f"Height: {bst.height()}")
    print(f"Total nodes: {bst.count_nodes()}")

    print(f"\nSearch 40: {'Found' if bst.search(40) else 'Not found'}")
    print(f"Search 99: {'Found' if bst.search(99) else 'Not found'}")

    print("\nDeleting 30 (node with two children)...")
    bst.delete(30)
    bst.display()
    print(f"Inorder after delete: {', '.join(map(str, bst.inorder()))}")

    print("\nDeleting 20 (node with one child)...")
    bst.delete(20)
    bst.display()

    print("\nDeleting 80 (leaf node)...")
    bst.delete(80)
    bst.display()
    print(f"Inorder after deletes: {', '.join(map(str, bst.inorder()))}")
